Adding dfx deps documentation page (#1708)
dfinity/portal: docs/developer-docs/setup/pulling-canister-dependencies.md

If a service canister also depends on other canisters, those dependencies should also be `pullable`. 

#### Non-`pullable` examples:

If the canister is meant for personal use and not intended for others to use, it does not make sense for the canister to be `pullable`. 

If a canister wasm is published for other developers to use for deploying their own instance, it does not make sense for the canister to be `pullable` since the canister ID of the instance is not static; users can test integrations locally and deploy them directly. An example of this canister type is the asset canister generated by dfx. 

## `Pullable` dfx.json example

For a canister to be `pullable`, the `dfx.json` file must include a `pullable` definition. Below is an example:

```json
{
  "canisters": {
    "service": {
      "type": "motoko",
      "main": "src/pullable/main.mo",
      "pullable": {
        "dependencies": [],
        "wasm_url": "https://github.com/lwshang/pullable/releases/latest/download/service.wasm",
        "init_guide": "A natural number, e.g. 1"
      }
    }
  }
}
```

:::caution

> The wasm module of a `pullable` canister must be hosted via an URL so that service consumers can download it when pulling the dependency.

Reviewer: I wonder if we should also add a footnote indicating that a future evolution of this feature will likely leverage direct wasm downloads from the replica.

`https://github.com/<USERNAME>/<REPONAME>/releases/latest/download/<FILENAME>`
:::

## Service provider workflow overview
First, a service provider must configure a canister to be `pullable` by setting it as such in the `dfx.json` file. 

An example of a provider `dfx.json` which has a `pullable` "service" canister can be found below:

```json
{
    "canisters": {
        "service": {
            "type": "motoko",
            "main": "src/main.mo",
            "pullable": {
                "wasm_url": "http://example.com/a.wasm",
                "wasm_hash": "d180f1e232bafcee7d4879d8a2260ee7bcf9a20c241468d0e9cf4aa15ef8f312",
                "dependencies": [
                    "yofga-2qaaa-aaaaa-aabsq-cai"
                ],
                "init_guide": "A natural number, e.g. 10."
            }
        }
    }
}
```

The `pullable` object will be serialized as a part of the `dfx` metadata and attached to the wasm.

To better understand the `pullable` object, let's look at each property in depth. 

- `wasm_url`: A URL used to download the canister wasm module which will be deployed locally.

> - `wasm_hash` A SHA256 hash of the wasm module located at `wasm_url`. This field is optional. In most cases, the wasm module at `wasm_url` will be the same as the on-chain wasm module. This means that dfx can read the state tree to obtain and verify the module hash. In other cases, the wasm module at `wasm_url` is not the same as the on-chain wasm module. For example, the Internet Identity canister provides a `development` flavor to be integrated locally. In these cases, `wasm_hash` provides the expected hash, and `dfx` verifies the downloaded wasm against this.

Reviewer: ```suggestion
- `wasm_hash` A SHA256 hash of the wasm module located at `wasm_url`. This field is optional. In most cases, the wasm module at `wasm_url` will be the same as the on-chain wasm module. This means that dfx can read the state tree to obtain and verify the module hash. In some cases, the wasm module at `wasm_url` is not the same as the on-chain wasm module. For example, the Internet Identity canister provides a `development` variant to be integrated locally. In these cases, `wasm_hash` provides the expected hash, and `dfx` verifies the downloaded wasm against this.
```

I also think we should expand on the implication of not receiving the expected hash. Will dfx print a warning? Should developers be concerned? When should they ignore the warning, and in what circumstances should they heed the warning?

Author: Updated this bullet point with information from @lwshang via Slack DM conversation.

- `dependencies`: An array of Canister IDs (`Principal`) of direct dependencies.
- `init_guide`: A message to guide consumers how to initialize the canister.

### Canister metadata requirements

A service provider canister used in production or in a production environment running on the mainnet should have public `dfx` metadata.

The canister wasm downloaded from `wasm_url` should have the following metadata (public or private):

- `candid:service`
- `candid:args`
- `dfx`

All metadata sections are handled by `dfx` when the canister is built.

### Deployment process

> Service providers will to use the following deployment process to deploy their `pullable` canister.

Reviewer: ```suggestion
Service providers will use the following deployment process to deploy their `pullable` canister.
```

- #### Step 1: First, assure that you have installed the [IC SDK](https://github.com/dfinity/sdk) and [git](https://git-scm.com/downloads).

- #### Step 2: Then, open a terminal window and use `git clone` to download the following repository:

```
git clone https://github.com/lwshang/pullable.git
```

- #### Step 3: Open the `dfx.json` file in a text or code editor. Declare the pull dependency in `dfx.json` file with the following configuration:

```
{
  "canisters": {
    "service": {
      "type": "pull",
      "id": "ig5e5-aqaaa-aaaan-qdxya-cai"
    }
  },
}
```

Save the file.

- #### Step 4: Use `dfx deps pull` to pull from the mainnet:

```
dfx deps pull
Fetching dependencies of canister ig5e5-aqaaa-aaaan-qdxya-cai...
Found 1 dependencies:
ig5e5-aqaaa-aaaan-qdxya-cai
Pulling canister ig5e5-aqaaa-aaaan-qdxya-cai...
```

- #### Step 5: Configure the `init` argument:

```
dfx deps init service --argument 1
```

- #### Step 6: Deploy on a local replica with the commands:

```
dfx start --clean --background
dfx deps deploy
dfx deploy app
```

- #### Step 7: Test the app by making the following call:
  
```

> dfx canister call app double_service

Reviewer: a little confusing here. why call `app` and not `service`? Does `app` in turn call `service`? Can we explain what is happening here?

Author: Updated this step with clarification from @lwshang via Slack DM conversation.

Installing canister: yofga-2qaaa-aaaaa-aabsq-cai
```

```
dfx deps deploy dep_b
Installing canister: yhgn4-myaaa-aaaaa-aabta-cai (dep_b)
```

:::info
`dfx deps deploy` always creates the canister with the anonymous identity so that dependencies and application canisters will have different controllers. It also always installs the canister in "reinstall" mode so that the canister status will be discarded.
:::

## Interactive example

Now that we've explored the concepts and overview of using the `dfx deps` workflow, let's take a look at using an interactive example to demonstrate the functionality. 

This example project will demonstrate an application canister pulling its dependency from the IC mainnet and integrating with it locally. 

In this example, the `app` canister defines a method called `double_service` which makes an inter-canister call to the `service` canister.

- #### Step 1: First, assure that you have installed the [IC SDK](https://github.com/dfinity/sdk) and [git](https://git-scm.com/downloads).

- #### Step 2: Then, open a terminal window and use `git clone` to download the following repository:

```
git clone https://github.com/lwshang/pullable.git
```

> 

Reviewer 2: Clone the service provider's repo is not a required step.

In this demo, a service consumer should `dfx new` to get new project and edit the `dfx.json` as in step 3 below.
